### DISCUSSION TITLE
Reducing the friction of supporting, maintaining, and contributing to Octokit

### DISCUSSION THREAD

**Discussion by nickfloyd** [octokit/.github#18] [Announcements]
Hey Octokit friends,

This will probably be one of the least glamorous things to discuss; hopefully, it will increase the quality of life around supporting, maintaining, and contributing to Octokit git a bit better.

We've been looking for ways to reduce the friction around contributing and supporting Octokit for a while now.  We can do several little things to increase our ability to service the roughly 70 repos that we currently own - one of those things is to drive toward consistency while ensuring each community still embraces its uniqueness around platform and language.  

You might've come across our discussions around generative SDKs ([here](https://github.com/octokit/octokit.net/discussions/2499),  [here](https://github.com/octokit/octokit.net/discussions/2527), and [here](https://github.com/octokit/octokit.net/discussions/2495#discussioncomment-3173935)) and how that will equip all of us to get more out of Octokit in a faster, more predictable way... Octokit.js already does this in many ways 🎉.

In the next few weeks, you'll probably experience some shuffling around some **meta aspects** (labels, PR and Issue templates, etc...) of the [Octokit](https://github.com/octokit) and [Terraform](https://github.com/integrations/terraform-provider-github) repos.

We want to be sensitive and avoid unnecessary workflows and over-expressed processes - so as you see things change, please "mind the dust" and know that we are iterating on the "right" things for our community. We're hoping to do our best to avoid the whole first-order thinking reaction of tearing things down (see [Chesterson's fence](https://en.wiktionary.org/wiki/Chesterton%27s_fence)) and defaulting to a position of second-order thinking where we consider the intent an purpose behind what is already in place.

As an example of the need mentioned above, have a look at the current "picture" of how we label things below - having all of these various ways to label things not only cause challenges with shipping but, as you know, they can be stumbling blocks for both new and old contributors 😬  we don't want that.

Thank you all for making this one of the most welcoming communities in open source.  **If you have any thoughts or feelings about things you've seen work in other open-source projects (around process, consistency, etc...), please add them as part of this discussion so our community can chat about them.**

Also, take time to call out any workflow or CI gotchas that could be problematic - we'll be going through these with a fine-tooth comb, but I am sure we'll probably miss some things.

![Screen Shot 2022-10-25 at 10 17 57 AM](https://user-images.githubusercontent.com/139819/197814562-829ceaa3-2402-4596-923f-8cec1902085f.png)

**Comment by kfcampbell** [comment]
I'd like to offer a couple of opinions regarding labels in hopes of prompting more discussion here.

1. We should remove many labels and shoot to keep the smallest possible subset of useful labels. 
1. Proposal for labels to keep:
    - Bug
    - Breaking change
    - Feature
    - Duplicate
    - Docs update only
    - Good first issue
    - Something to signify "blocked on upstream API support"
    - Something to signify "housekeeping/chore/dependency update"
    - A set of standardized sizing labels
1. I think ideally we'd standardize on a cadence of merging and releasing small dependency updates as well as a proscribed timeframe to respond to CVEs.

**Reply by gr2m** [reply]
> * We should remove many labels and shoot to keep the smallest possible subset of useful labels.

Yes
 
> * Proposal for labels to keep:
>   
>   * Bug
>   * Breaking change
>   * Feature
>   * Duplicate
>   * Docs update only
>   * Good first issue
>   * Something to signify "blocked on upstream API support"
>   * Something to signify "housekeeping/chore/dependency update"
>   * A set of standardized sizing labels

Other Labels that I think worked well for the JS repositories in the past:
- Maintenance
   - this is what we use for "housekeeping/chore/dependency update"
- Documentation
- Blocked
   - I think we should have a general "blocked" label, and then clarify with either sub-labels or an explanation in the issue

I think it's also helpful to have some language specific labels in order to group issues. For JS repositories that's `typescript` and `deno`. I'd like to add `esm`. 

Another label that could be relevant for grouping across languages is "Enterprise". Other labels might be "REST" and "GraphQL", but we don't have a need for these two in JavaScript Octokit due to the multi-repo setup.

Lastly, I have a simple GitHub App that enforces a category label: "Bug", "Feature", "Maintenance". I wanted to add "Documentation" as another category label. The app also adds the issues based on the category labels to a [project](https://github.com/orgs/octokit/projects/1) which I used extensively with my GitHub Manager while I was still a contracter on Octokit. Today I'd use the new Projects.

> * I think ideally we'd standardize on a cadence of merging and releasing small dependency updates as well as a proscribed timeframe to respond to CVEs.

No strong opinion. For dependency updates I'd set the goal for no more than one week, at least for JS.

**Reply by timrogers** [reply]
Agree with @gr2m that we should allow the possibility of repo-/language-specific repos alongside the shared ones that we use across Octokit.

**Comment by nickfloyd** [comment]
Here is the current working set of labels: https://github.com/nickfloyd/repo-meta/labels

Here are the suggested ISSUE templates: https://github.com/nickfloyd/repo-meta/tree/main/.github/ISSUE_TEMPLATE

Try out the templates: https://github.com/nickfloyd/repo-meta/issues/new/choose

**Reply by kfcampbell** [reply]
Nice! Some quick thoughts: 

- I think the labels look good! I'd be happy to use those and iterate as necessary. 
- Issue templates
    - One thing to consider for `bug.yml` would be something like "reported impact of bug": what is this blocking? 
    - Why is REST API version needed in each of these templates? Do each of our SDKs allow the user to call multiple REST API versions within an SDK version? 
    - Is "relevant log output" needed for `feature.yml`?
    - `pull_request.md` (should be `.yml`?) is rendering weird for me (picture below). Also, it may be worth considering adding a "breaking change" checkbox to that template. 
![image](https://user-images.githubusercontent.com/9327688/198106484-a1de9275-94a8-4171-ae62-705979160a02.png)

**Reply by gr2m** [reply]
I'd give all groups of labels the same color. And all labels that don't belong to a group I'd just make grey

**Reply by gr2m** [reply]
instead of "REST API Version", we could ask for GitHub Version, as in: github.com, GHES [VERSION], GHEC? Besides REST, APIs also include GraphQL, Webhooks, and Authentication/Authorization

**Reply by timrogers** [reply]
I think we should have both 😊

**Reply by gr2m** [reply]
Ahh wait, with REST API Version you mean the new not-yet-released versioning, right? Sorry I missed that at first. 

There is a risk asking for too much stuff, but yeah, we could ask for both. It's another case where e.g. in `github/graphql.js` I would not ask for the REST API Version though, for example

**Reply by kfcampbell** [reply]
I really like the idea of GitHub versions! There's a decent amount of API surface differences between the various GitHub flavors.

**Comment by timrogers** [comment]
I think we should try, if possible, to avoid the word "chore". It devalues work that people (and us) are doing and it would be better to use more descriptive tags like "dependencies" or "docs".

**Reply by nickfloyd** [reply]
Agreed! I ended up going with `Maintenance`

**Comment by timrogers** [comment]
I'm not sure what we should use `Status: Invalid` as people might find that judgemental. I'd probably suggest just closing an issue or PR instead.

**Reply by nickfloyd** [reply]
Yeah, I was on the fence about this one. In some of the SDKs, I have reserved it for when I get spammy stuff.

**Reply by nickfloyd** [reply]
How about: [Status: Invalid/Incomplete](https://github.com/nickfloyd/repo-meta/labels/Status%3A%20Invalid%2FIncomplete)?

**Reply by timrogers** [reply]
How about "Needs more information"?

**Reply by gr2m** [reply]
I use `invalid` exclusively for spammy issues. `spam` might actually be a better label, more explicit, at least for how I use it.

**Reply by timrogers** [reply]
Spam is okay for the most egregious cases. I just don't want to upset people if they're trying their best!

**Reply by gr2m** [reply]
> I just don't want to upset people if they're trying their best!

in that case I just wouldn't add a special label. I'd close it as "won't fix". Usually people file a bug or feature request, then we realize it's really a support request, so I remove `bug`/`feature`, add `support`, and close the issue when the discussion concluded

**Reply by timrogers** [reply]
I guess it depends on what we're trying to get to. Nick's mention of "incomplete" made me think that he was looking for a state when the issue does not have enough information.

**Comment by timrogers** [comment]
I'm not sure that "Good first issue" should be a "type" - I'd expect an issue to have a single type, and then this would be a "tag".

**Reply by kfcampbell** [reply]
Agreed! In my original comment I did not distinguish between "type" and "tag".

**Comment by timrogers** [comment]
For the JavaScript repos, when submitting bugs, I think ideally we would ask people to provide all of the Octokit package versions from their `package-lock.json` file (or similar). We often have issues that are related to particular combinations of versions.

**Reply by nickfloyd** [reply]
Good point... though I think that would be better encapsulated in PR or Issue content and not labels.

**Reply by timrogers** [reply]
Yes. This would be for the issue templates.

**Comment by gr2m** [comment]
Regarding issue templates: I'd like to adjust them based on the needs of the repositories. I think we can agree on a group of fields that should be present and use the same names for aggregation, but then we should be able to adjust descriptions and add additional fields etc as needed.

**Comment by gr2m** [comment]
Another category label we use a lot is "Support". I guess we could move these issues to discussions though. 

What happens a lot is that folks open a bug report or feature request, and then we realize it's not actually a bug or the feature already exists or there is a better way to achieve what they want. In this case I change the category label from `bug`/`feature` to `support`.

What I don't like about transferring to discussions is that most often these issues can be concluded, while discussions are open ended.

**Reply by timrogers** [reply]
"Support" is a good point!

**Reply by nickfloyd** [reply]
Would support and question overlap in that case? If so, I think I'd prefer support over having both.

**Reply by kfcampbell** [reply]
I like the idea of taking support over question.

**Reply by gr2m** [reply]
yes, I never needed question

**Comment by nickfloyd** [comment]
We have implemented the changes as discussed across all Octokit and Terraform SDK repos.
